src/warnings.ts

```diff
@@ -0,0 +1,156 @@
+    this._options = Object.assign({}, defaultFilterOptions, this._options);
+  }
+
+  shouldIgnore(warning: Warning) {
+    if (this._options.warningCodesToIgnore.has(warning.code)) {
+      return true;
+    }
+    if (warning.severity > this._options.minimumSeverity) {
+      return true;
+    }
+    return false;
+  }
+}
+
+const defaultPrinterOptions = {
+  verbosity: Verbosity.Full,
+  color: true
+};
+
+export class WarningPrinter {
+  _chalk: typeof chalk;
+  constructor(
+      private _outStream: NodeJS.WritableStream,
+      private _options?: PrinterOptions) {
+    this._options = Object.assign({}, defaultPrinterOptions, _options);
+    this._chalk =
+        new (chalk.constructor as any)({enabled: this._options.color});
+  }
+
+  printWarnings(warnings: Iterable<Warning>) {
+    for (const warning of warnings) {
+      this.printWarning(warning);
+    }
+  }
+
+  printWarning(warning: Warning) {
+    const severity = this._severityToString(warning.severity);
+    const range = warning.sourceRange;
+
+    if (this._options.verbosity >= Verbosity.Full) {
+      this._outStream.write('\n');
+      const lineText = this._getTextOfLine(range.start.line, range.file);
+      this._outStream.write(`${lineText}\n`);
+      const colorFunction = this._severityToColorFunction(warning.severity);
+      const underlineText = getUnderlineText(lineText, range);
+      this._outStream.write(`${colorFunction(underlineText)}\n`);
+      this._outStream.write('\n');
+    }
+
+    this._outStream.write(
+        `${range.file}` +
+        `(${range.start.line},${range.start.column}) ` +
+        `${severity} [${warning.code}] - ${warning.message}\n`);
+  }
+
+  private _severityToString(severity: Severity) {
+    const colorFunction = this._severityToColorFunction(severity);
+    switch (severity) {
+      case Severity.ERROR:
+        return colorFunction('error');
+      case Severity.WARNING:
+        return colorFunction('warning');
+      case Severity.INFO:
+        return colorFunction('info');
+      default:
+        const never: never = severity;
+        throw new Error(`Unknown severity value - ${never
+                        } - encountered while printing warning.`);
+    }
+  }
+
+  private _severityToColorFunction(severity: Severity) {
+    switch (severity) {
+      case Severity.ERROR:
+        return this._chalk.red;
+      case Severity.WARNING:
+        return this._chalk.yellow;
+      case Severity.INFO:
+        return this._chalk.green;
+      default:
+        const never: never = severity;
+        throw new Error(`Unknown severity value - ${never
+                        } - encountered while printing warning.`);
+    }
+  }
+  private _getTextOfLine(line: number, localPath: string) {
+    const contents =
+        fs.readFileSync(path.join(this._options.basedir, localPath), 'utf-8');
+    return contents.split('\n')[line];
+  }
+}
+
+function repeatCharacter(char: string, num: number) {
+  const chars: string[] = [];
```

Comment:
Clever!

(`num + 1` as it turns out, but I didn't catch that either until the tests failed)